src/wasm/wasm-type.cpp

```diff
@@ -96,39 +92,25 @@ void Type::init(const std::vector<Type>& types) {
   }
 #endif
 
-  if (types.size() >= (1 << SIZE_BITS)) {
-    WASM_UNREACHABLE("Type too large");
+  if (types.size() == 0) {
+    id = none;
+    return;
   }
-  _size = types.size();
-
-  auto lookup = [&]() {
-    auto indexIt = indices.find(types);
-    if (indexIt != indices.end()) {
-      id = indexIt->second;
-      return true;
-    } else {
-      return false;
-    }
-  };
-
-  {
-    // Try to look up previously interned type
-    std::shared_lock<std::shared_timed_mutex> lock(mutex);
-    if (lookup()) {
-      return;
-    }
+  if (types.size() == 1) {
+    *this = types[0];
+    return;
   }
```

Comment:
I think I don't understand how this all works yet (maybe some overview in types.h?). Here in `init` we take a lock on anything that doesn't have 0 or 1 elements. Are we somehow not calling init as much? If so I missed that part (or something else).

Reply:
The problem was never that we called `init` a lot, it was that we called `expand` a lot and it acquired the lock. Now `expand` never acquires the lock. I'll add documentation about how the IDs work in the header.

Reply:
Right, thanks. Now I get it!